File: src/commands/list.ts
import { Command } from 'commander';
import chalk from 'chalk';
import Table from 'cli-table3';
import { InstallerService } from '../services/installer';
import { RegistryService } from '../services/registry';

export const listCommand = new Command('list')
  .description('List available or installed agents')
  .option('-i, --installed', 'list installed agents only')
  .option('-u, --updates', 'list agents with available updates')
  .option('-c, --category <category>', 'filter by category')
  .option('--json', 'output as JSON')
  .action(async (options) => {
    try {
      const installer = new InstallerService(options.registry);
      const registry = new RegistryService(options.registry);

      if (options.installed) {
        await listInstalledAgents(installer, options);
      } else if (options.updates) {
        await listUpdates(installer, options);
      } else {
        await listAvailableAgents(registry, options);
      }
    } catch (error) {
      console.error(chalk.red('Error listing agents:'), error instanceof Error ? error.message : String(error));
      process.exit(1);
    }
  });

async function listInstalledAgents(installer: InstallerService, options: any) {
  const installed = await installer.getInstalledAgents();
  
  if (installed.length === 0) {
    console.log(chalk.yellow('No agents installed.'));
    return;
  }

  if (options.json) {
    console.log(JSON.stringify(installed, null, 2));
    return;
  }

  const table = new Table({
    head: [
      chalk.cyan('ID'),
      chalk.cyan('Version'),
      chalk.cyan('Target'),
      chalk.cyan('Installed'),
    ],
    colWidths: [25, 12, 15, 20],
  });

  for (const agent of installed) {
    table.push([
      agent.id,
      agent.version,
      agent.target,
      new Date(agent.installedAt).toLocaleDateString(),
    ]);
  }

  console.log(table.toString());
  console.log(chalk.gray(`\n${installed.length} agent(s) installed`));
}

async function listUpdates(installer: InstallerService, options: any) {
  const updates = await installer.checkForUpdates();
  
  if (updates.length === 0) {
    console.log(chalk.green('All agents are up to date!'));
    return;
  }

  if (options.json) {
    console.log(JSON.stringify(updates, null, 2));
    return;
  }

  const table = new Table({
    head: [
      chalk.cyan('ID'),
      chalk.cyan('Current'),
      chalk.cyan('Latest'),
      chalk.cyan('Target'),
    ],
    colWidths: [25, 12, 12, 15],
  });

  for (const { agent, latestVersion } of updates) {
    table.push([
      agent.id,
      agent.version,
      chalk.green(latestVersion),
      agent.target,
    ]);
  }

  console.log(table.toString());
  console.log(chalk.yellow(`\n${updates.length} update(s) available`));
  console.log(chalk.gray('Run "agents update" to update all agents'));
}

async function listAvailableAgents(registry: RegistryService, options: any) {
  const filters = {
    category: options.category,
    limit: 50, // Default limit for available agents
  };

  const agents = await registry.searchAgents('', filters);
  
  if (agents.length === 0) {
    console.log(chalk.yellow('No agents found.'));
    return;
  }

  if (options.json) {
    console.log(JSON.stringify(agents, null, 2));
    return;
  }

  const table = new Table({
    head: [
      chalk.cyan('ID'),
      chalk.cyan('Name'),
      chalk.cyan('Author'),
      chalk.cyan('Category'),
      chalk.cyan('Downloads'),
      chalk.cyan('Rating'),
    ],
    colWidths: [20, 25, 15, 15, 12, 10],
  });

  for (const agent of agents) {
    table.push([
      agent.id,
      agent.name.en,
      agent.author,
      agent.category,
      agent.downloads.toLocaleString(),
      `${agent.rating.toFixed(1)}/5`,
    ]);
  }

  console.log(table.toString());
  console.log(chalk.gray(`\nShowing ${agents.length} agent(s)`));
}

File: cmd/list.go
package cmd

import (
	"fmt"
	"os"
	"text/tabwriter"

	"github.com/spf13/cobra"
	"github.com/chameleon-nexus/agents-cli/pkg/registry"
	"github.com/chameleon-nexus/agents-cli/pkg/installer"
)

var listCmd = &cobra.Command{
	Use:   "list",
	Short: "List available or installed agents",
	Long: `List agents from the registry or show locally installed agents.

Examples:
  agents list                    # List all available agents
  agents list --installed        # List installed agents
  agents list --category=development  # List by category
  agents list --updates          # List agents with updates`,
	RunE: func(cmd *cobra.Command, args []string) error {
		// Get flags
		installed, _ := cmd.Flags().GetBool("installed")
		category, _ := cmd.Flags().GetString("category")
		updates, _ := cmd.Flags().GetBool("updates")

		if installed {
			return listInstalledAgents()
		} else if updates {
			return listUpdatableAgents()
		} else {
			return listAvailableAgents(category)
		}
	},
}

func init() {
	// Add flags
	listCmd.Flags().Bool("installed", false, "List installed agents")
	listCmd.Flags().String("category", "", "Filter by category")
	listCmd.Flags().Bool("updates", false, "List agents with available updates")
}

func listAvailableAgents(category string) error {
	client := registry.NewClient()
	
	agents, err := client.GetAllAgents()
	if err != nil {
		return fmt.Errorf("failed to get agents: %w", err)
	}

	// Filter by category if specified
	if category != "" {
		var filtered []registry.Agent
		for _, agent := range agents {
			if agent.Category == category {
				filtered = append(filtered, agent)
			}
		}
		agents = filtered
	}

	if len(agents) == 0 {
		fmt.Println("No agents found")
		return nil
	}

	// Create table writer
	w := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
	fmt.Fprintln(w, "NAME\tVERSION\tCATEGORY\tDOWNLOADS\tRATING\tDESCRIPTION")
	fmt.Fprintln(w, "----\t-------\t--------\t---------\t------\t-----------")

	for _, agent := range agents {
		fmt.Fprintf(w, "%s\t%s\t%s\t%d\t%.1f\t%s\n",
			agent.ID,
			agent.Latest,
			agent.Category,
			agent.Downloads,
			agent.Rating,
			truncateString(agent.Description.En, 50))
	}

	w.Flush()
	fmt.Printf("\nTotal: %d agents\n", len(agents))
	return nil
}

func listInstalledAgents() error {
	inst := installer.NewClaudeCodeInstaller()
	
	agents, err := inst.ListInstalled()
	if err != nil {
		return fmt.Errorf("failed to list installed agents: %w", err)
	}

	if len(agents) == 0 {
		fmt.Println("No agents installed")
		return nil
	}

	// Create table writer
	w := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
	fmt.Fprintln(w, "NAME\tVERSION\tINSTALLED\tPATH")
	fmt.Fprintln(w, "----\t-------\t---------\t----")

	for _, agent := range agents {
		fmt.Fprintf(w, "%s\t%s\t%s\t%s\n",
			agent.ID,
			agent.Version,
			agent.InstalledAt.Format("2006-01-02"),
			agent.Path)
	}

	w.Flush()
	fmt.Printf("\nTotal: %d installed agents\n", len(agents))
	return nil
}

func listUpdatableAgents() error {
	client := registry.NewClient()
	inst := installer.NewClaudeCodeInstaller()
	
	installed, err := inst.ListInstalled()
	if err != nil {
		return fmt.Errorf("failed to list installed agents: %w", err)
	}

	if len(installed) == 0 {
		fmt.Println("No agents installed")
		return nil
	}

	var updatable []installer.InstalledAgent
	
	for _, agent := range installed {
		metadata, err := client.GetAgentMetadata(agent.ID)
		if err != nil {
			continue // Skip if can't get metadata
		}
		
		if metadata.Latest != agent.Version {
			updatable = append(updatable, agent)
		}
	}

	if len(updatable) == 0 {
		fmt.Println("All agents are up to date")
		return nil
	}

	// Create table writer
	w := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
	fmt.Fprintln(w, "NAME\tCURRENT\tLATEST\tSTATUS")
	fmt.Fprintln(w, "----\t-------\t------\t------")

	for _, agent := range updatable {
		metadata, _ := client.GetAgentMetadata(agent.ID)
		fmt.Fprintf(w, "%s\t%s\t%s\tUpdate Available\n",
			agent.ID,
			agent.Version,
			metadata.Latest)
	}

	w.Flush()
	fmt.Printf("\nTotal: %d agents can be updated\n", len(updatable))
	fmt.Println("Run 'agents update' to update all agents")
	return nil
}

func truncateString(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen-3] + "..."
}
